strawberryfields/apps/train/parametrization.py

```diff
@@ -0,0 +1,406 @@
+# Copyright 2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""
+Contains the :class:`~.VGBS` (variational GBS) class which provides a trainable parametrization
+of the GBS probability distribution.
+"""
+from typing import Optional
+
+import numpy as np
+import thewalrus.samples
+from thewalrus._hafnian import reduction
+from thewalrus._torontonian import tor
+from thewalrus.quantum import Qmat, Xmat
+from thewalrus.quantum import find_scaling_adjacency_matrix as rescale
+from thewalrus.quantum import find_scaling_adjacency_matrix_torontonian as rescale_tor
+from thewalrus.quantum import photon_number_mean_vector, pure_state_amplitude
+
+import strawberryfields as sf
+
+
+def rescale_adjacency(A: np.ndarray, n_mean: float, threshold: bool) -> np.ndarray:
+    """Rescale an adjacency matrix so that it can be mapped to GBS.
+
+    An adjacency matrix must have singular values not exceeding one if it can be mapped to GBS.
+    Arbitrary adjacency matrices must first be rescaled to satisfy this condition.
+
+    This function rescales an input adjacency matrix :math:`A` so that the corresponding gaussian
+    state has:
+
+    - a mean number of *clicks* equal to ``n_mean`` when ``threshold=True``;
+    - a mean number of *photons* equal to ``n_mean`` when ``threshold=False``.
+
+    **Example usage:**
+
+    >>> a = np.ones((3, 3))
+    >>> rescale_adjacency(a, 2, True)
+    array([[0.32232919, 0.32232919, 0.32232919],
+           [0.32232919, 0.32232919, 0.32232919],
+           [0.32232919, 0.32232919, 0.32232919]])
+
+    Args:
+        A (array): the adjacency matrix to rescale
+        n_mean (float): the target mean number of clicks or mean number of photons
+        threshold (bool): determines whether rescaling is for a target mean number of clicks or
+            photons
+
+    Returns:
+        array: the rescaled adjacency matrix
+    """
+    scale = rescale_tor(A, n_mean) if threshold else rescale(A, n_mean)
+    return A * scale
+
+
+def _Omat(A: np.ndarray) -> np.ndarray:
+    """Find the :math:`O` matrix of an adjacency matrix.
+
+    Args:
+        A (array): the adjacency matrix
+
+    Returns:
+        array: the :math:`O` matrix of :math:`A`
+    """
+    n = len(A)
+    A_big = np.block([[A, 0 * A], [0 * A, np.conj(A)]])
```

Comment:
(Minor suggestion): would this be better for readability? (let me know if I missed something)
```suggestion
    A_big = np.block([[A, np.zeros_like(A)], [np.zeros_like(A), np.conj(A)]])
```

Reply:
Done